### PR TITLE
[HZ-384] Generate the attribution files during the build

### PR DESCRIPTION
**Pull request by kwart**
This PR adds `attribution.txt` files to our artifacts (within the `META-INF`) and also an aggregated one to the `licenses` directory in the ZIP/TGZ distributions.

A new `attribution-maven-plugin` is used for creating the attribution files: https://github.com/hazelcast/attribution-maven-plugin

Example of the attribution file format:
```
biz.aQute:bndlib:1.50.0
	Copyright (c) 2000 Gatespace AB. All Rights Reserved.
	Copyright (c) OSGi Alliance (2002, 2006, 2007). All Rights Reserved.
	Copyright (c) OSGi Alliance (2009, 2010). All Rights Reserved.
	Copyright (c) OSGi Alliance (2011). All Rights Reserved.

com.amazonaws:aws-java-sdk-core:1.11.976
	Copyright (c) 2016 Amazon.com, Inc. or its affiliates. All Rights Reserved.
	Copyright (c) 2016-2019. Amazon.com, Inc. or its affiliates. All Rights Reserved.
	Copyright (c) 2016. Amazon.com, Inc. or its affiliates. All Rights Reserved.
	Copyright (c) 2019. Amazon.com, Inc. or its affiliates. All Rights Reserved.
	Copyright 2010-2021 Amazon.com, Inc. or its affiliates. All Rights
	Copyright 2010-2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
	Copyright 2011-2021 Amazon Technologies, Inc.
	Copyright 2011-2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
	Copyright 2012-2021 Amazon Technologies, Inc.
	Copyright 2012-2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
	Copyright 2013-2021 Amazon Technologies, Inc.
	Copyright 2013-2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
	Copyright 2014-2021 Amazon Technologies, Inc.
	Copyright 2014-2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
	Copyright 2015-2021 Amazon Technologies, Inc.
	Copyright 2015-2021 Amazon.com, Inc. or its affiliates. All Rights
	Copyright 2015-2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
	Copyright 2016-2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
	Copyright 2018-2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
	Copyright 2019-2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
	Copyright 2020-2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
	Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.

...
```

### CODE REVIEW

**Comment by kwart** [comment]
run-lab-run

File: pom.xml [hunk 1]
```diff
@@ -528,6 +529,32 @@
                 </executions>
                 <inherited>false</inherited>
             </plugin>
+            <plugin>
+                <groupId>com.hazelcast.maven</groupId>
+                <artifactId>attribution-maven-plugin</artifactId>
```
Comment (frant-hartm) [inline pom.xml:534]: Where does this plugin come from?

File: pom.xml [hunk 2]
```diff
@@ -587,6 +614,22 @@
...
+                    <configuration>
+                        <exclusionPatterns>
+                            <exclusionPattern>^Copyright Laws of the United States.</exclusionPattern>
+                            <exclusionPattern>per-property override</exclusionPattern>
+                            <exclusionPattern>^copyright notice</exclusionPattern>
+                            <exclusionPattern>FAILED_PRECONDITION</exclusionPattern>
+                            <exclusionPattern>The ASF licenses this file to you under the Apache License</exclusionPattern>
+                            <exclusionPattern>Substitute for l, l</exclusionPattern>
+                            <exclusionPattern>inside edge crossings that are between</exclusionPattern>
+                        </exclusionPatterns>
```
Comment (frant-hartm) [inline pom.xml:630]: Maybe some explanation for this would be good.

**Review by frant-hartm — CHANGES_REQUESTED** [review]

File: pom.xml (hunk 1 above)
Comment (kwart) [reply pom.xml:534]: This is a new home-grown plugin that greps the dependency sources and searches for the "Copyright ..." lines there.
https://github.com/hazelcast/attribution-maven-plugin

File: pom.xml (hunk 2 above)
Comment (kwart) [reply pom.xml:630]: :+1: I'll add an explanation.

**Review by olukas — APPROVED** [review]

**Review by frant-hartm — APPROVED** [review]

**Comment by kwart** [comment]
Thanks for reviews, Ondra & Franta.

**Pull request merged**